modules/statistics-service-api/src/test/java/org/opencastproject/statistics/api/StatisticsUtilTest.java

```diff
@@ -49,6 +49,31 @@ private static Instant randomInstant() {
     return Instant.ofEpochSecond(randomEpochSec);
   }
 
+  @Test
+  public void testGetBucketsHourly() {
+    for (int i = 0; i < repetitions / 24; i++) {
+      final Instant a = randomInstant();
+      final Instant b = randomInstant();
+      final Instant from = Ordering.natural().min(a, b);
+      final Instant to = Ordering.natural().max(a, b);
+      logger.info("from {} to {} {}", from, to, DataResolution.HOURLY.name());
+      final List<Instant> buckets = StatisticsUtil.getBuckets(from, to, DataResolution.HOURLY, ZoneId.of("Z"));
+      assertEquals(from, buckets.get(0));
+      if (buckets.size() == 1) {
+        continue;
+      }
+      assertEquals(
+          LocalDateTime.ofInstant(
+              to,
+              ZoneOffset.UTC)
+              .withMinute(0)
+              .withSecond(0)
+              .withNano(0)
+              .toInstant(ZoneOffset.UTC),
+          buckets.get(buckets.size() - 1));
+    }
```

Comment:
We also check the first bucket, but not the ones in between :) Using the random Instants was easier than sitting down an thinking about all possible edge cases. 

So the answer to both of your questions is: laziness (or "prioritization" in other words :)). I will spend some extra time to improve the tests.

Reply:
I have extended the tests a bit.

Reply:
Alright, now I'm wondering why we need the randomised tests at all if we now have a nice set of predefined instants to test. I'm not too fond of random test cases because this could lead to tests failing only sometimes, which is terrible for debugging...

Reply:
Well, I wouldn't consider 3 date pairs a "nice set of predefined instants to test". The idea with the random tests is to log the input so you can debug it very easily. However, I just noticed that Opencast has set the log level to "WARN" for all tests running with maven so my logs are don't get printed. Therefore, I added the from and to dates to the output which is printed in case a test fails.

I actually like randomized testdata since most of the time, people just test the cases they can imagine (which are mostly identical with the ones they have in mind when developing).

Though it still makes sense to have hand-picked test data as well to make sure that some important cases are always tested.